### ISSUE TITLE
Discrepancies between model parameters and periodogram

### ISSUE THREAD

**Issue by johnh2o2** [PrincetonUniversity/FastTemplatePeriodogram#21]
Hi Jake,

I've added a branch that I have been using to update the API for the `FastTemplateModeler` class; in the course of unit testing, I've uncovered some discrepancies which need to be addressed and understood before moving forward. I think these issues are related to the discrepancies you were finding between `SlowTemplateModeler` and `FastTemplateModeler` (i.e. `SlowTemplateModeler` was able to find better solutions at some frequencies).

## Setup:

These issues are recreated by running the `custom_testing.py` script (which generates an animation illustrating several hopefully-related problems). One of the changes I've made in this branch is the addition of a `TemplateModel` class, which contains the model fit (frequency, template, and template fit parameters (`ModelFitParams`)), and is callable. 

Here's a smaller example that uses the `TemplateModel` to inject a signal and to produce the model fit `y_fit` from the `best_fit_pars` returned by `fit_template`.

```python
from pyftp.modeler import TemplateModel
from pyftp.template import Template
from pyftp.utils import ModelFitParams, weights
from pyftp.fast_template_periodogram import fit_template
import numpy as np

ndata = 100
frequency = 1.0
sigma = 0.1

c_n=[0.1, 0.5, 0.1]
s_n=[0.5, 0.5, 0.1]
a, b, c, sgn = 1.0, 1.0, 0.0, 1

# generate template
template = Template(c_n=c_n, s_n=s_n)

# set up model parameters
parameters = ModelFitParams(a=a, b=b, c=c, sgn=sgn)

model = TemplateModel(template, frequency=frequency, parameters=parameters)

# generate random observation times
t = np.sort(np.random.rand(ndata))

# generate signal
y = model(t)

# inject white noise
y += sigma * np.random.randn(ndata)
yerr = sigma * np.ones_like(y)

# recover parameters
p_max, best_fit_pars = fit_template(t, y, yerr, template, frequency, allow_negative_amplitudes=True)

# obtain model fit
model_fit = TemplateModel(template, frequency=frequency, parameters=best_fit_pars)
y_fit = model_fit(t)


# use definition of periodogram to compute periodogram from best_fit_pars and signal
w = weights(yerr)
ybar = np.dot(w, y)
chi2_0 = np.dot(w, (y - ybar)**2) 
chi2_model = np.dot(w, (y - y_fit)**2)
chi2_signal = np.dot(w, (y - model(t))**2)

p_model = 1 - chi2_model / chi2_0
p_signal = 1 - chi2_signal / chi2_0

print("p_max (fit_template) = {p_max}\np_model (computed from best_fit_pars) = {p_model}\np_signal (computed from signal parameters) = {p_signal}".format(p_max=p_max, p_model=p_model, p_signal=p_signal))
```

1. If you run this script, you should see that `p_max`, `p_model` and `p_signal` **all disagree**
2. If you run the `custom_testing.py` script, you should see an animation that shows
      a. The model as defined by `best_fit_pars` compared with the true signal and noisy data
      b. Three periodogram values: one for the periodogram value returned by `fit_template`, one for the true underlying signal, and one for the periodogram implied by `best_fit_pars` (though this is not visible due to being << 1). 
      c. The difference between `freq * tau_best_fit - freq * tau_signal` vs `freq * tau_signal`.

![last_frame_h1](https://cloud.githubusercontent.com/assets/5678551/24053354/29dc04ac-0b0f-11e7-98d2-083512ec61b2.png)

![last_frame_h3](https://cloud.githubusercontent.com/assets/5678551/24053426/6a3a8cd0-0b0f-11e7-968c-36a1ca82498b.png)
The above is a snapshot of the animation at the last frame (H=1 case top, H=3 case bottom)

## Observations:
1. The periodogram value returned by `fit_template` is inconsistent with the best fit parameters returned by `fit_template`. 
2. The periodogram value returned by `fit_template` is sub-optimal for certain values of `tau` (around `freq * tau` = 0, 0.5, 1)
3. There seems to be a systematic phase offset between the best fit parameters and the true signal. This is **close to** a phase shift of 0.5, but the periodogram for a phase-shifted model is still not consistent with the periodogram reported by `fit_template` nor the "true" periodogram value. (shown below).
4. @jakevdp  you were mentioning that you found instances where the `SlowTemplateModeler` was able to find more accurate solutions than the `FastTemplateModeler`, which I believe is related to this collection of problems.

![last_frame_h3_phi0_0 50](https://cloud.githubusercontent.com/assets/5678551/24053958/00ec0d24-0b11-11e7-91b9-cebb1db7f797.png)

Let me know if you have any thoughts about this. I'll keep working on this in the meantime!

John

**Comment by jakevdp** [comment]
Interesting – thanks for putting this together.

Given how complex the various routines are that lead to this, I think the most fruitful way to track down the bug will be to make sure each component is well tested. My suspicion from playing around with the nfft is that there may be corner cases where the fast summations do not accuratly approximate the true summations – if we could add an option to build the periodogram using direct summations, that could help.

It might also be some kind of interaction between errors in root finding and errors in summation approximation. But I wouldn't dive into those kind of second order effects until each component is understood and validated by comprehensive unit tests.

**Comment by johnh2o2** [comment]
Hi Jake,

Thanks for the feedback -- the computations above (which use the new `fit_template` function) only use `direct_summations`, so this is independent of the NFFT. I suspect there may be some numerical instabilities somewhere that cause the problem at certain phase shifts. Root finding from computed coefficients is known to be unstable with certain algorithms and with certain classes of polynomials (closely spaced roots, large variations in coefficient amplitudes, etc), so that may explain some of the problem.

However, I'm still currently baffled by why the reported fit parameters should be inconsistent with the reported periodogram value, since they both are computed from the same set of variables, and the periodogram computation depends explicitly on the phase shift. 

I'll keep adding tests and digging deeper. I'm working on reviewing a JOSS article at the moment, so I'll get back to this a bit later tonight, but feel free to poke around if you have any ideas.

John

**Comment by johnh2o2** [comment]
Another note -- this branch implements unit tests that compare `direct_summations` against `fast_summations`, and everything passes. Doesn't have any bearing on this problem (I don't think), but that should give some additional piece of mind about the NFFT operations. And, since they're unit tested, you can run amok with any speedup ideas for the `fast_summations` function if you feel like it.

**Comment by jakevdp** [comment]
Ah, I missed that part. Good to know.

**Comment by johnh2o2** [comment]
Long story short, I tracked this down to a couple of issues. 

* One was a mistake in the `PseudoPolynomial.root_finding_poly()` method. Originally, the method would return `(1 - x^2)p^2(x) - q^2(x)`, however, this should be `p^2(x) - (1 - x^2)q^2(x)`.

* The phase shift parameter, which has caused much grief, was also incorrect, and this was a result of the affine transformation of the observation times to the `(-1/2, 1/2)` range that was required by NFFT. I had hacked together what I thought was a fix, but this should really be taken care of by the `fast_summations` function. Now, the results of the NFFT are multiplied by a set of `twiddle` factors, which correct for the transformation.

Everything should now be fixed as of commit `b30e775ef1684c64d3e5574dce30bd7ad8d6241a`.
